users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -106,13 +119,66 @@ def add_specaug_layer(
     # Update config with pre-generated parameters
     full_config["specaug_params"] = specaug_params
 
+
     network[name] = {
         "class": "eval",
         "from": from_list,
         "eval": f"self.network.get_config().typed_value('transform')(source(0, as_data=True), network=self.network, **{full_config})",
     }
+    return [name], [
+        sort_filters_by_center_freq,
+        get_frequency_response,
+        filter_based_masking,
+        _mask,
+        random_mask,
+        transform,
+    ]
+
+
+def filter_based_masking(x, batch_axis, axis, probability_distribution, max_number_masks_for_probability_specaug):
+    """
+    :param tf.Tensor x: (batch,time,feature)
+    :param int batch_axis:
+    :param int axis:
+    :param tf.Tensor probability_distribution:
+    :param int max_number_masks_for_probability_specaug:
+    """
+    import tensorflow as tf
+
+    n_batch = tf.shape(x)[batch_axis]  # Batch size
+    ndim = x.get_shape().ndims  # Number of dimensions in x
+    n_features = tf.shape(x)[axis]  # Feature dimension size
+
+    # Random number of masks for each batch  shape: (n_batch,)
+    num_masks = tf.random.uniform(
+        shape=[n_batch], minval=0, maxval=max_number_masks_for_probability_specaug + 1, dtype=tf.int32
+    )
+    # Total number of masks to generate across all batches
+    total_masks = tf.reduce_sum(num_masks)
+
+    # Sample feature indices to be masked based on the probability distribution (categorical needs 2d) shape: (1, total_masks)
+    masked_features = tf.random.categorical(tf.expand_dims(probability_distribution, 0), total_masks)
+    # Generate batch offsets to map the masks to the corresponding batches. example: [0,0,1,2,2,2]
+    batch_indices = tf.repeat(tf.range(n_batch), num_masks)
 
-    return [name], get_specaug_funcs()
+    # Gather the selected feature indices (have to match the shape)
+    feature_indices = tf.gather(tf.reshape(masked_features, [-1]), tf.range(total_masks))  # shape: (total_masks,)
+    feature_indices = tf.reshape(tf.cast(feature_indices, tf.int32), [-1])  # shape: (total_masks,)
+
+    # Combine batch offsets and feature indices
+    indices = tf.stack([batch_indices, feature_indices], axis=1) # shape: (total_masks, 2)
+
+    # Create a boolean mask initialized to False. shape: (n_batch, n_features)
+    mask = tf.tensor_scatter_nd_update(
+        tf.zeros((n_batch, n_features), dtype=tf.bool), indices, tf.ones_like(indices[:, 0], dtype=tf.bool)
+    )
+    # Reshape the mask to match the input tensor shape
+    mask_shape = [tf.shape(x)[i] if i in (batch_axis, axis) else 1 for i in range(ndim)] # shape: (batch, time, feature)
+    mask = tf.reshape(mask, mask_shape)                                          # shape: (batch, time, feature)
+
+    x = tf.where(mask, tf.zeros_like(x), x)
+
+    return x
 
 
 def transform(data, network, **config):
```

Comment:
I would add this new version as `transform_with_filter_variance_masking` and leave the existing `transform` function the same. Otherwise, currently/previously running experiments are not reproducible anymore.